torch/csrc/autograd/variable.h

```diff
@@ -449,17 +449,28 @@ struct TORCH_API DifferentiableViewMeta : public AutogradMeta {
 // Differentiable view. Track history with DifferentiableViewMeta.
 inline Variable make_variable_differentiable_view(
     Variable base,
-    at::Tensor data,
+    at::Tensor& data,
     CreationMeta creation_meta,
     c10::optional<std::function<at::Tensor(const at::Tensor&)>> view_func = c10::nullopt) {
   if (data.defined()) {
-    auto data_impl_copy = data.getIntrusivePtr()->shallow_copy_and_detach(
-      /*version_counter=*/0,
-      /*allow_tensor_metadata_change=*/true);
-    data_impl_copy->set_autograd_meta(std::make_unique<DifferentiableViewMeta>(
-      data_impl_copy.get(), std::move(base), std::move(view_func),
-      creation_meta));
-    return Variable(data_impl_copy);
+    // If we already did a TensorImpl allocation for data, just reuse it.
+    // Otherwise(e.g tensor.swapdim(0, 0) when we return the same tensor as input),
+    // we have to use shallow_copy_and_detach to create a new TensorImpl to avoid
+    // moving leaf node into graph interior. This guarantees only 1 TensorImpl
+    // allocation happens in view ops.
+    if (data.getIntrusivePtr().use_count() == 1 && data.getIntrusivePtr()->unique_version()) {
```

Comment:
Btw, you changed this due to my comment, but I'm not convinced this version is better than the old one. In particular, you now have to do an atomic load to get the use count; in the previous implementation where you just compared with self, there were no atomics involved. This won't show up in instruction counts but it if there is contention the atomic will matter.

Reply:
Yea I agree but I checked the change still gives similar perf gain as the direct comparison of tensorimpls. 
One thing I like about `use_count()` is here we make sure we are the only user before we modify it. While in the old implementation (although currently we don't have other users) we might be changing the Tensor while other "users" don't notice. 
I feel the changes are easier to understand than before.